### PR TITLE
1813: URLs printed by Skara should have ".git" at the end of the repo name

### PR DESCRIPTION
**Pull request by zhaosongzs**
In this patch, `.git` is added to the URLs printed by Skara.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1813](https://bugs.openjdk.org/browse/SKARA-1813): URLs printed by Skara should have ".git" at the end of the repo name

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [95d2c6e2](https://git.openjdk.org/skara/pull/1467/files/95d2c6e27c1f2f3c8acd7d6f3a8077b2fcda9b5b)
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - Committer) ⚠️ Review applies to [959eb93b](https://git.openjdk.org/skara/pull/1467/files/959eb93b4f0fdf831669d81c215ce03df0658947)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara pull/1467/head:pull/1467` \
`$ git checkout pull/1467`

Update a local copy of the PR: \
`$ git checkout pull/1467` \
`$ git pull https://git.openjdk.org/skara pull/1467/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1467`

View PR using the GUI difftool: \
`$ git pr show -t 1467`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1467.diff">https://git.openjdk.org/skara/pull/1467.diff</a>

</details>

### CODE REVIEW

**Review by erikj79 — CHANGES_REQUESTED** [review]
I don't think we should just add ".git" to the webUrl. What we need is a new method on `HostedRepository` that returns the URL in a format suitable for the command line client. Basically the form you get in the "Code" drop down on the repository main page on GitHub. There is already a method `HostedRepository::url` which is almost this but not quite. It doesn't perform the transformation, and it adds authentication, which we don't want. Maybe `HostedRepository::transformedUrl`, not sure.

**Comment by zhaosongzs** [comment]
> I don't think we should just add ".git" to the webUrl. What we need is a new method on `HostedRepository` that returns the URL in a format suitable for the command line client. Basically the form you get in the "Code" drop down on the repository main page on GitHub. There is already a method `HostedRepository::url` which is almost this but not quite. It doesn't perform the transformation, and it adds authentication, which we don't want. Maybe `HostedRepository::transformedUrl`, not sure.

Ok, I will add a new method. Thanks.

File: forge/src/main/java/org/openjdk/skara/forge/HostedRepository.java [hunk 1]
```diff
@@ -83,6 +83,7 @@ PullRequest createPullRequest(HostedRepository target,
     URI webUrl(String baseRef, String headRef);
     URI diffUrl(String prId);
     VCS repositoryType();
+    URI remoteUrl();
```
Comment (erikj79) [inline forge/src/main/java/org/openjdk/skara/forge/HostedRepository.java:86]:
```suggestion
    /**
     * Returns a URL suitable for CLI interactions with the repository
     */
    URI remoteUrl();
```

**Review by erikj79 — COMMENTED** [review]
This is better. There are more places that need to use the new method. The fallback instructions for /backport is one example where I just saw URLs with no .git in example command lines.

**Comment by zhaosongzs** [comment]
> This is better. There are more places that need to use the new method. The fallback instructions for /backport is one example where I just saw URLs with no .git in example command lines.

I would use it in suitable places.

File: bots/checkout/src/main/java/org/openjdk/skara/bots/checkout/CheckoutBot.java [hunk 2]
```diff
@@ -64,10 +64,7 @@ private static String urlEncode(URI uri) {
     }
 
     private URI webURI() {
-        var webURI = from.webUrl().toString();
-        if (!webURI.endsWith(".git")) {
-            webURI += ".git";
-        }
+        var webURI = from.remoteUrl().toString();
```
Comment (erikj79) [inline bots/checkout/src/main/java/org/openjdk/skara/bots/checkout/CheckoutBot.java:67]: Converting from URI to String and then back to URI doesn't seem like a good idea. This method is called in one place, so should probably just get rid of it.

Comment (zhaosongzs) [reply bots/checkout/src/main/java/org/openjdk/skara/bots/checkout/CheckoutBot.java:67]: Oh, I didn't find this redundant operation.

File: bots/checkout/src/main/java/org/openjdk/skara/bots/checkout/CheckoutBot.java [hunk 3]
```diff
@@ -64,12 +64,7 @@ private static String urlEncode(URI uri) {
     }
 
     private URI webURI() {
-        var webURI = from.webUrl().toString();
-        if (!webURI.endsWith(".git")) {
-            webURI += ".git";
-        }
-
-        return URI.create(webURI);
+        return from.remoteUrl();
```
Comment (erikj79) [inline bots/checkout/src/main/java/org/openjdk/skara/bots/checkout/CheckoutBot.java:67]: I meant that this whole method is redundant now and can just be inlined.

Comment (zhaosongzs) [reply bots/checkout/src/main/java/org/openjdk/skara/bots/checkout/CheckoutBot.java]: Sorry, fixed it now.

**Review by lgxbslgx — APPROVED** [review]
Looks good. The name `remoteUrl` doesn't seem very good, but with the comment, it is OK.

**Comment by erikj79** [comment]
> Looks good. The name `remoteUrl` doesn't seem very good, but with the comment, it is OK.

I agree, it's hard to come up with a good name for this. Ideally I think this one should be just `url` and the existing `url` should be renamed something like `authenticatedUrl`.

**Comment by zhaosongzs** [comment]
> > Looks good. The name `remoteUrl` doesn't seem very good, but with the comment, it is OK.
> 
> I agree, it's hard to come up with a good name for this. Ideally I think this one should be just `url` and the existing `url` should be renamed something like `authenticatedUrl`.

Yes, I found it difficult to come up with a name at that time. I will rename the methods.

**Comment by zhaosongzs** [comment]
This pr is still waiting for review.

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java [hunk 4]
```diff
@@ -1883,8 +1883,8 @@ void testIssueOriginalRepo(TestInfo testInfo) throws IOException {
             var comment = comments.get(0);
             assertTrue(comment.body().contains(editHash.toString()));
             // Verify that the 'original' repo URL is used in the comment and not the main one
-            assertTrue(comment.body().contains(originalRepo.url().toString()));
-            assertFalse(comment.body().contains(repo.url().toString()));
+            assertTrue(comment.body().contains(originalRepo.authenticatedUrl().toString()));
+            assertFalse(comment.body().contains(repo.authenticatedUrl().toString()));
```
Comment (erikj79) [inline bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:1887]: I think we should be checking for `.url()` here.

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java [hunk 5]
```diff
@@ -1986,7 +1986,7 @@ void testAltFixVersionsMatch(TestInfo testInfo) throws IOException {
             assertEquals(1, comments.size());
             var comment = comments.get(0);
             assertTrue(comment.body().contains(editHash.toString()));
-            assertTrue(comment.body().contains(repo.url().toString()));
+            assertTrue(comment.body().contains(repo.authenticatedUrl().toString()));
```
Comment (erikj79) [inline bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:1989]: Same here.

File: bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java [hunk 6]
```diff
@@ -2034,7 +2034,7 @@ void testAltFixVersionsMatchRegex(TestInfo testInfo) throws IOException {
...
```
Comment (erikj79) [inline bots/notify/src/test/java/org/openjdk/skara/bots/notify/issue/IssueNotifierTests.java:2037]: And again.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java [hunk 7]
```diff
@@ -1213,7 +1213,7 @@ private void checkStatus() {
 
             if (pr.sourceRepository().isPresent()) {
                 var branchNames = pr.repository().branches().stream().map(HostedBranch::name).collect(Collectors.toSet());
-                if (!pr.repository().url().equals(pr.sourceRepository().get().url()) && branchNames.contains(pr.sourceRef())) {
+                if (!pr.repository().authenticatedUrl().equals(pr.sourceRepository().get().authenticatedUrl()) && branchNames.contains(pr.sourceRef())) {
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:1216]: I think we should be comparing `.url()` here. If we compare `authenticatedUrl()` there is a risk that different user authentications would trigger a false negative.

File: bots/pr/src/main/java/org/openjdk/skara/bots/pr/LimitedCensusInstance.java [hunk 8]
```diff
@@ -115,7 +114,7 @@ private static Optional<JCheckConfiguration> jCheckConfiguration(
     }
 
     private static Path getRepoFolder(HostedRepositoryPool hostedRepositoryPool, HostedRepository censusRepo, String censusRef, Path folder) {
-        var repoName = censusRepo.url().getHost() + "/" + censusRepo.name();
+        var repoName = censusRepo.authenticatedUrl().getHost() + "/" + censusRepo.name();
```
Comment (erikj79) [inline bots/pr/src/main/java/org/openjdk/skara/bots/pr/LimitedCensusInstance.java:117]: Only need `.url()` to extract host.

File: bots/tester/src/test/java/org/openjdk/skara/bots/tester/InMemoryHostedRepository.java [hunk 9]
```diff
@@ -234,6 +234,11 @@ public URI webUrl(Tag tag) {
         return null;
     }
 
+    @Override
+    public URI url() {
+        return null;
```
Comment (erikj79) [inline bots/tester/src/test/java/org/openjdk/skara/bots/tester/InMemoryHostedRepository.java:239]: It probably doesn't matter much, but seems like this should return `url` just like `authenticatedUrl()` does.

File: cli/src/main/java/org/openjdk/skara/cli/pr/GitPrCreate.java [hunk 10]
```diff
@@ -305,7 +305,7 @@ public static void main(String[] args) throws IOException, InterruptedException
         }
 
         var mailingLists = new ArrayList<String>();
-        var parentProject = ForgeUtils.projectName(parentRepo.url());
+        var parentProject = ForgeUtils.projectName(parentRepo.authenticatedUrl());
```
Comment (erikj79) [inline cli/src/main/java/org/openjdk/skara/cli/pr/GitPrCreate.java:308]: Should probably be `.url()`.

File: forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java [hunk 11]
```diff
@@ -530,7 +530,7 @@ private String metadataFor(Check check) {
     }
 
     private String linkToDiff(String path, Hash hash, int line) {
-        return "[" + path + " line " + line + "](" + URIBuilder.base(repository.url())
+        return "[" + path + " line " + line + "](" + URIBuilder.base(repository.authenticatedUrl())
```
Comment (erikj79) [inline forge/src/main/java/org/openjdk/skara/forge/gitlab/GitLabMergeRequest.java:533]: This looks like another case of `.url()`. The authentication is even removed below. This string is built to be presented as text somewhere.

File: test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java [hunk 12]
```diff
@@ -56,7 +56,7 @@ public TestHostedRepository(TestHost host, String projectName, Repository localR
         this.host = host;
         this.projectName = projectName;
         this.localRepository = localRepository;
-        pullRequestPattern = Pattern.compile(url().toString() + "/pr/" + "(\\d+)");
+        pullRequestPattern = Pattern.compile(authenticatedUrl().toString() + "/pr/" + "(\\d+)");
```
Comment (erikj79) [inline test/src/main/java/org/openjdk/skara/test/TestHostedRepository.java:59]: Doesn't matter much as they resolve to the same value, but logically I think `.webUrl()` fits better here.

File: test/src/main/java/org/openjdk/skara/test/TestPullRequest.java [hunk 13]
```diff
@@ -214,7 +214,7 @@ public boolean isDraft() {
     @Override
     public URI webUrl() {
         try {
-            return new URI(targetRepository.url().toString() + "/pr/" + id());
+            return new URI(targetRepository.authenticatedUrl().toString() + "/pr/" + id());
```
Comment (erikj79) [inline test/src/main/java/org/openjdk/skara/test/TestPullRequest.java:217]: Logically this should be `.webUrl()`.

**Review by erikj79 — COMMENTED** [review]
This turned into a big patch. I looked through all the `url()` -> `authenticatedUrl()` changes and found some that should be reverted or changed in some way.

File: bots/hgbridge/src/main/java/org/openjdk/skara/bots/hgbridge/JBridgeBot.java [hunk 14]
```diff
@@ -120,16 +120,16 @@ public Collection<WorkItem> run(Path scratchPath) {
 
             IOException lastException = null;
             for (var destination : exporterConfig.destinations()) {
-                var markerBase = destination.url().getHost() + "/" + destination.name();
+                var markerBase = destination.authenticatedUrl().getHost() + "/" + destination.name();
```
Comment (zhaosongzs) [inline bots/hgbridge/src/main/java/org/openjdk/skara/bots/hgbridge/JBridgeBot.java:123]: I think here should be `.url()` too.

**Comment by zhaosongzs** [comment]
@erikj79 Thank you very much for reviewing this big patch and identifying some issues!

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**